### PR TITLE
enhance: [C172] Utility functions unit test

### PR DESCRIPTION
**Pull request by tienhoah**
- Add unit tests for routeUtils
- Expand tests for mapUtils and chartUtils

## Every PR

Before merging, the developer of this pull request has checked the following:

- [x] Changes have been tested locally without errors
- [x] Lint has run and any errors have been resolved
- [x] Prettier has run on any changed files
- [x] Storybook stories have run and any errors have been resolved
- [x] Pre-existing unit tests have run and passed
  - [x] Unit tests have been added or updated, where possible, to prevent future regressions

## Post-Merge

- [ ] Code has merged and build success is confirmed in 'Actions'
- [ ] The corresponding Trello ticket has moved to the appropriate list (likely User-QA)

## Release Notes

* **Tests**
  * Added comprehensive unit test coverage for chart utility functions including region labeling, chart data generation, and plume data mapping.
  * Expanded map utility tests covering benthic styling, geometric bounds, map interaction management, zonal statistics, and feature querying.
  * Introduced new test suite for route parameter validation and normalization.

* **Chores**
  * Enhanced TypeScript type safety in map utilities.

### CODE REVIEW

File: src/tests/chartUtils.test.ts [hunk 1]
```diff
@@ -469,3 +478,290 @@ describe('export filename utilities', () => {
...
+    yAxisTitle: 'chart_information.watersheds',
+    legendColors: { w1: '#aabbcc', w2: '#bbccdd', w3: '#ccddee' },
+    width: 0.8,
+    barmode: 'stack',
+  },
+}
+
+// ---------------------------------------------------------------------------
+// getRegionLabel
+// ---------------------------------------------------------------------------
```
Comment (JBurkinshaw) [inline src/tests/chartUtils.test.ts]: These comments are probably overkill given the function name is in the argument of the the describe function. Remove them if they're not adding any context

File: src/tests/chartUtils.test.ts [hunk 2]
```diff
@@ -469,3 +478,290 @@ describe('export filename utilities', () => {
...
+    expect(getUpOneLevelLabel('country', mockRegion)).toBe('regions.global')
+  })
+})
+
+// ---------------------------------------------------------------------------
+// buildChartDataFromProperties
+// ---------------------------------------------------------------------------
+describe('buildChartDataFromProperties', () => {
+  beforeEach(() => {
+    jest.clearAllMocks()
```
Comment (JBurkinshaw) [inline src/tests/chartUtils.test.ts]: I don't think any of the 4 describe blocks creates a persistent mock to clear. `setChartData = jest.fn()` is created fresh inside each `it()`, and `i18next.t = ...` is a direct assignment, not a jest mock. `jest.clearAllMocks()` doesn't do anything here and can be deleted here and elswhere

File: src/tests/mapUtils.test.ts [hunk 3]
```diff
@@ -294,4 +325,500 @@ describe('map utilities', () => {
...
+      )
+      expect(sandIdx).toBeGreaterThan(-1)
+      expect(expr[sandIdx + 1]).toBe(colors['sand'])
+    })
+  })
+
+  describe('postZonalStats', () => {
+    it('returns parsed JSON on a successful response', async () => {
+      const mockData = { result: 'ok' }
+      jest.spyOn(global, 'fetch').mockResolvedValue({
```
Comment (JBurkinshaw) [inline src/tests/mapUtils.test.ts:494]: Several tests do `jest.spyOn(global, 'fetch').mockResolvedValue(...)` inline, but nothing ever calls `mockRestore` / `restoreAllMocks`. After this suite runs, `global.fetch` is left mocked. Inside this file each test replaces the previous mock so it works in isolation, but it could pollute elswhere makes failures hard to attribute.

To fix, add `afterEach(() => jest.restoreAllMocks())` to the surrounding describe (or use `mockImplementationOnce` so behaviour reverts naturally).

File: src/tests/chartUtils.test.ts [hunk 4]
```diff
@@ -469,3 +478,290 @@ describe('export filename utilities', () => {
...
+    expect(getRegionLabel('country', mockRegion as never, null)).toBe('Pacific Region')
+
+    expect(getRegionLabel('region', mockRegion as never, null)).toBe('Pacific Region')
+  })
+})
+
+// ---------------------------------------------------------------------------
+// getDrawerTitle
+// ---------------------------------------------------------------------------
+describe('getDrawerTitle', () => {
```
Comment (JBurkinshaw) [inline src/tests/chartUtils.test.ts:542]: `getDrawerTitle` -  the function itself is just several if statements. The tests pretty much restate the implementation - if a future dev changes `'global_trends'` to `'global.trends'` the test only catches the same typo the function would. Either drop the test or condense the branches into a single `it.each` table to make the spec readable at a glance.

The same applies for `getRegionLabel`, `getEffectiveRegionType` and `getUpOneLevelLabel`

File: src/tests/mapUtils.test.ts [hunk 5]
```diff
@@ -294,4 +325,500 @@ describe('map utilities', () => {
...
+
+      const bounds = calculateFeatureBounds(feature)
+      expect(bounds._sw.lng).toBeCloseTo(178.0)
+      expect(bounds._sw.lat).toBeCloseTo(-18.0)
+      expect(bounds._ne.lng).toBeCloseTo(179.0)
+      expect(bounds._ne.lat).toBeCloseTo(-17.0)
+    })
+  })
+
+  describe('clearPolygonHover', () => {
```
Comment (JBurkinshaw) [inline src/tests/mapUtils.test.ts]: There are 6 tests for three trivial setters around `map.setFeatureState`. One combined "polygon hover/select state helpers" test with a couple of assertions would be ok here.

File: src/tests/routeUtils.test.ts [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    it('returns parsed float for a mid-range zoom value', () => {
+      expect(getValidZoom('10.5')).toBe(10.5)
+    })
+
+    it('returns null for empty string', () => {
+      expect(getValidZoom('')).toBeNull()
+    })
+  })
+
+  describe('getValidLatLng', () => {
```
Comment (JBurkinshaw) [inline src/tests/routeUtils.test.ts:152]: There are 13 cases, with lots of overlap (one test per boundary instead of `it.each`). The `accepts boundary lat values` and `accepts boundary lng values` cases also overlap with `returns parsed floats when both lat and lng are valid`.

File: src/tests/routeUtils.test.ts [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+      expect(getValidLatLng('0', '-180')).toEqual({ lat: 0, lng: -180 })
+      expect(getValidLatLng('0', '180')).toEqual({ lat: 0, lng: 180 })
+    })
+
+    it('accepts negative valid coordinates', () => {
+      expect(getValidLatLng('-17.5', '-178.5')).toEqual({ lat: -17.5, lng: -178.5 })
+    })
+  })
+
+  describe('getValidDispersalPoint', () => {
```
Comment (JBurkinshaw) [inline src/tests/routeUtils.test.ts:182]: There are 11 tests here, but the function delegates entirely to `getValidLatLng` after splitting. Three tests cover what's unique here: null/empty input, wrong number of comma-separated parts, valid round-trip. The rest just retest `getValidLatLng` so probably aren't necessary

File: src/tests/routeUtils.test.ts [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    it('returns null for whitespace-only string', () => {
+      expect(getValidWatershed('   ')).toBeNull()
+    })
+
+    it('trims leading and trailing whitespace from valid input', () => {
+      expect(getValidWatershed('  watershed-123  ')).toBe('watershed-123')
+    })
+  })
+
+  describe('getValidYear', () => {
```
Comment (JBurkinshaw) [inline src/tests/routeUtils.test.ts:68]: The `returns parsed year when it is in availableYears` test (just 2020) and `returns parsed year for each available year` test (all five) overlap

File: src/utils/mapUtils.ts [hunk 9]
```diff
@@ -426,18 +426,22 @@ export async function postZonalStats(payload) {
...
     2005: SEDIMENT_EXPOSURE_2005_URL,
     2010: SEDIMENT_EXPOSURE_2010_URL,
     2015: SEDIMENT_EXPOSURE_2015_URL,
     2020: SEDIMENT_EXPOSURE_2020_URL,
   }
 
+  const resolvedUrl = exposureUrls[year]
+  if (!resolvedUrl) {
+    throw new Error(`No sediment exposure URL available for year: ${year}`)
+  }
```
Comment (JBurkinshaw) [inline src/utils/mapUtils.ts:440]: Replacing the silent `exposureUrls[year]` lookup with a typed map and an explicit `throw` is a real improvement.
Previously, an unmapped year would have quietly sent `url: undefined` to the API. Just worth confirming callers handle the new throw: `getAllYearZonalStats` swallows it via `catch { return { [year]: {} } }`, which is fine, but any direct caller will now surface an error where they used to get `undefined`. Worth a quick scan of the call sites to make sure nothing relies on the old silent-failure behaviour.

**Review by JBurkinshaw — CHANGES_REQUESTED** [review]
It's great to see the test coverage increase. But increased coverage is a balance because every test is also a maintenance liability. Tests that only restate the implementation tend to fail and pass at the same time as the code they cover - that can give false confidence and slows down future refactors (you end up updating the test alongside the code, rather than the test catching a regression). In my opinion, the worthwhile tests are the ones that test specific behaviour and edge cases. Places where the logic might not be obvious or easy to break. Worth aiming for fewer, sharper tests rather than one per branch. I have added some suggestions for you in the code.

Not for this PR, (I will create a ticket). The app is Vite-based and Vitest is already a dependency for Storybook tests, so running Jest for unit tests means two test runners, two configs, and two mental models for what should be one concern. `setupTests.ts` already has a workaround that exists purely because Jest doesn't understand Vite's `import.meta.env`, which Vitest handles natively by sharing Vite's transform pipeline. The migration cost here is low - three test files plus setup, and the API is near-identical (mostly a `jest.*` → `vi.*` rename) - so a separate PR that adds a Vitest unit-test project alongside the existing Storybook one, drops the Jest dependencies, and removes the `import.meta` hack would be a good simplification. Keeping it isolated from feature work makes review and revert easy.

Also, not for this PR which just follows the pattern already established by the existing `mapChartConfigToData` tests at the top of the file. 
`chartUtils.test.ts:176, 203, 250, 299` (and pre-existing usage above)
```ts
Object.assign(chartSeriesConfig, sharedLandUseMockConfig)
```
We should follow up on this: `Object.assign` permanently overwrites entries on the imported `chartSeriesConfig` for the rest of the test process, with no `afterEach` to restore. Anything else that imports `chartSeriesConfig` (this file's other tests, other files) sees the mutated values, so we're relying on test ordering not changing.

Fixes could include snapshot/restore in `beforeEach`/`afterEach`, use `vi.doMock`/`jest.doMock` on `../data/chartSeriesData`, or pass `chartSeriesConfig` in as a parameter to the functions under test (cleanest long-term).

File: src/tests/chartUtils.test.ts (hunk 2 above)
Comment (tienhoah) [reply src/tests/chartUtils.test.ts]: Thanks Joe, I remove them. A common AI mistake is attempting to create redundant code.

File: src/tests/mapUtils.test.ts (hunk 3 above)
Comment (tienhoah) [reply src/tests/mapUtils.test.ts:494]: @JBurkinshaw 
We have `restoreMocks: true` in `jest.config.js`, which is Jest's global equivalent of calling `jest.restoreAllMocks()` after every single test automatically. So `global.fetch` is already being restored after each test, and no manual `afterEach` is needed.

File: src/utils/mapUtils.ts (hunk 9 above)
Comment (tienhoah) [reply src/utils/mapUtils.ts:440]: Thanks Joe, the only caller in app code is `getAllYearZonalStats`, which handles the error correctly. We are clear that nothing relied on the old silent `undefined` behaviour.

File: src/tests/routeUtils.test.ts (hunk 8 above)
Comment (tienhoah) [reply src/tests/routeUtils.test.ts:68]: Updated

File: src/tests/routeUtils.test.ts (hunk 7 above)
Comment (tienhoah) [reply src/tests/routeUtils.test.ts:182]: Updated

File: src/tests/routeUtils.test.ts (hunk 6 above)
Comment (tienhoah) [reply src/tests/routeUtils.test.ts:152]: Updated

File: src/tests/mapUtils.test.ts (hunk 5 above)
Comment (tienhoah) [reply src/tests/mapUtils.test.ts]: Updated

File: src/tests/chartUtils.test.ts (hunk 4 above)
Comment (tienhoah) [reply src/tests/chartUtils.test.ts:542]: Updated

File: src/tests/chartUtils.test.ts (hunk 1 above)
Comment (tienhoah) [reply src/tests/chartUtils.test.ts]: Updated

**Review by JBurkinshaw — APPROVED** [review]
Looks great. Much cleaner and more focused now
